### PR TITLE
Studio: Delete all active demo sites

### PR DESCRIPTION
**Pull request by kozer**
Related to https://github.com/Automattic/dotcom-forge/issues/6655
Fixes https://github.com/Automattic/dotcom-forge/issues/7081

## Proposed Changes

This PR fixes a bug where the user reaches the maximum number of active demo sites. However, if none of the created sites are linked to the demo sites, he can't remove any to continue creating demo sites. 

## Testing Instructions
Prerequisite: Apply D148749-code

1. Create some sites.
2. Edit `appdata-v1.json` files, and remove all snapshots
3. Restart the app
4. Go to settings, and ensure that you see a number of active demo sites like the below:
2024-05-15T09:36:58,113953464+03:00
5. Press the 3 dots menu, and then press the "Delete all demo sites" button.
6. Ensure that you see a message like below:
2024-05-15T09:37:57,606601158+03:00
7. Ensure that pressing cancel doesn't have any effect.
8. Repeat steps from 4-6.
9. On the dialog press "Delete all demo sites"
10. Notice that the sites are deleted, and you end up with 0 active demo sites.

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

**Comment by sejas** [comment]
@kozer , what do you think about moving the heavy logic to the backend? I feel more confident if the App makes only one request to delete all the sites.

File: src/hooks/use-site-usage.ts [hunk 1]
```diff
@@ -19,10 +24,11 @@ export function useSiteUsage() {
 		}
 		setIsLoading( true );
 		try {
-			const response: { site_count: number; site_limit: number } = await client.req.get( {
-				path: '/jurassic-ninja/usage',
-				apiNamespace: 'wpcom/v2',
-			} );
+			const response: { site_count: number; site_limit: number; sites: UsageSite[] } =
```
Comment (sejas) [inline src/hooks/use-site-usage.ts]: We could infer `site_count` from the number of `sites`.

**Comment by katinthehatsite** [comment]
@kozer it seems that you are adding a confirmation window for deleting all sites in this PR. I currently have a similar issue assigned: https://github.com/Automattic/dotcom-forge/issues/7081 - would you like to assign it to yourself if that case since you are covering that functionality?

File: src/hooks/use-site-usage.ts (hunk 1 above)
Comment (kozer) [reply src/hooks/use-site-usage.ts]: Yes, I know, however, I thought not to change API to keep consistency with users that haven't updated studio, and still have an older version.
Given that, and as the API already returns that information, I kept it as is :)

**Comment by kozer** [comment]
> @kozer it seems that you are adding a confirmation window for deleting all sites in this PR. I currently have a similar issue assigned: https://github.com/Automattic/dotcom-forge/issues/7081 - would you like to assign it to yourself if that case since you are covering that functionality?

I'm sorry, Kat! The ticket I'm working on, mentioned the confirmation in the comments, and I haven't realized, there was a separate issue for that! I can assign it to me, thanks!

**Comment by kozer** [comment]
> @kozer , what do you think about moving the heavy logic to the backend? I feel more confident if the App makes only one request to delete all the sites.

@sejas , I was thinking about that, and I decided not to do that, for the following reasons:
1. We have all that logic in the client already implemented.
2. It wouldn't change anything more, than just moving things around, as still, we will have to loop the sites and call `request_site_deletion`, on each one of them. Any of those delete requests can fail asynchronously ( as we call jobs in that function, that schedule a job, and call other apis ), so again we could have a network partition.
So given that, I thought that although seems strange to have multiple requests, I don't see the real value of moving all that complicated logic of handling failures etc, as firstly, it works perfectly as is, and I don't see the clear benefit by doing so.
What do you think?

**Comment by katinthehatsite** [comment]
> I'm sorry, Kat! The ticket I'm working on, mentioned the confirmation in the comments, and I haven't realized, there was a separate issue for that! I can assign it to me, thanks!

No worries, I have not started yet - so all good! I removed my assignment from that issue 👍

**Review by katinthehatsite — COMMENTED** [review]
Hey I am testing this PR now and once I delete all the snapshots and navigate to Settings > three dots and attempt to click on the button to delete all sites, I can't seem to make proceed further as the button appears blocked:

Capture d’écran, le 2024-05-16 à 09 37 42

Any ideas what might be happening?

**Comment by kozer** [comment]
> Hey I am testing this PR now and once I delete all the snapshots and navigate to Settings > three dots and attempt to click on the button to delete all sites, I can't seem to make proceed further as the button appears blocked:
> 
> Capture d’écran, le 2024-05-16 à 09 37 42
> Any ideas what might be happening?

Tbh, I had the same problem this morning, @katinthehatsite . I applied the diff, and I didn't have any success.
I restarted the proxy and worked as expected!

To see if it's working or not, open dev tools and check if `/usage` endpoint resolves. For me it didn't

File: src/components/user-settings.tsx [hunk 2]
```diff
@@ -162,11 +160,25 @@ export default function UserSettings() {
...
-	}, [ snapshots, deleteAllSnapshots ] );
+
+		const CANCEL_BUTTON_INDEX = 0;
+		const DELETE_BUTTON_INDEX = 1;
+
+		const { response } = await getIpcApi().showMessageBox( {
+			type: 'warning',
+			message: __( 'Delete all demo sites' ),
+			detail: __( "Any changes you've made to your demo sites will be lost." ),
+			buttons: [ __( 'Cancel' ), __( 'Delete all' ) ],
```
Comment (katinthehatsite) [inline src/components/user-settings.tsx:189]: should the `Delete all` be set to blue color similar to the native dialogue that we have for deleting a single site?

Screenshot 2024-05-16 at 10 53 54 AM

**Comment by katinthehatsite** [comment]
> Tbh, I had the same problem this morning, @katinthehatsite . I applied the diff, and I didn't have any success.
I restarted the proxy and worked as expected!
To see if it's working or not, open dev tools and check if /usage endpoint resolves. For me it didn't

I see, I did not apply the diff initially because I missed `Prerequisite: Apply D148749` 🤦‍♀️ However, now it works for me as expected when I follow the testing steps. I only have one comment on my end regarding buttons color.

File: src/components/user-settings.tsx (hunk 2 above)
Comment (matt-west) [reply src/components/user-settings.tsx:189]: If we have the ability to make this red, we should as it's a destructive action.

Comment (kozer) [reply src/components/user-settings.tsx:189]: > should the Delete all be set to blue color similar to the native dialogue that we have for deleting a single site?

I'm not sure I can do that @katinthehatsite . In my machine, there is no styling at all. Maybe can you help me here somehow?

Comment (wojtekn) [reply src/components/user-settings.tsx:189]: @kozer could we make the order of the buttons consistent for both confirmation dialogs?

Also, could we use the text from the existing dialog but modify it to account for multiple demo sites being removed, eg.?

> All demo sites' databases that exist for your WordPress.com account, along with all posts, pages, comments, and media, will be lost.

**Review by wojtekn — COMMENTED** [review]
When I tested the feature, I had some floating demo sites in Studio: a few expired ones, and one I removed using Jurassic Ninja Report Card tool. The new feature removed existing sites correctly, but it didn't clear the floating ones. Should it also remove all local demo sites?

**Comment by matt-west** [comment]
> Prerequisite: Apply [D148749-code](https://href.li/?https://code.a8c.com/D148749)

@kozer I don't have access to that diff to apply it.

**Comment by kozer** [comment]
> > Prerequisite: Apply [D148749-code](https://href.li/?https://code.a8c.com/D148749)
> 
> @kozer I don't have access to that diff to apply it.

It's ok Matt! I'm not sure what I need to do for you to obtain access. My guess is that @wojtekn can do that for you.

**Comment by sejas** [comment]
> > @kozer , what do you think about moving the heavy logic to the backend? I feel more confident if the App makes only one request to delete all the sites.
> 
> @sejas , I was thinking about that, and I decided not to do that, for the following reasons:
> 
> 1. We have all that logic in the client already implemented.
> 2. It wouldn't change anything more, than just moving things around, as still, we will have to loop the sites and call `request_site_deletion`, on each one of them. Any of those delete requests can fail asynchronously ( as we call jobs in that function, that schedule a job, and call other apis ), so again we could have a network partition.
>    So given that, I thought that although seems strange to have multiple requests, I don't see the real value of moving all that complicated logic of handling failures etc, as firstly, it works perfectly as is, and I don't see the clear benefit by doing so.
>    What do you think?

Thanks for sharing the context! Let's do that 👌

File: src/components/user-settings.tsx [hunk 3]
```diff
@@ -162,11 +162,27 @@ export default function UserSettings() {
...
-	}, [ snapshots, deleteAllSnapshots ] );
+
+		const CANCEL_BUTTON_INDEX = 0;
+		const DELETE_BUTTON_INDEX = 1;
+
+		const { response } = await getIpcApi().showMessageBox( {
+			type: 'warning',
+			message: __( 'Delete all demo sites' ),
+			detail: __(
+				'All demo sites databases that exist for your WordPress.com account, along with all posts, pages, comments, and media, will be lost.'
```
Comment (wojtekn) [inline src/components/user-settings.tsx]: Should we drop the `databases` word to make it clearer that demo sites are deleted, and not only databases?

```suggestion
				'All demo sites that exist for your WordPress.com account, along with all posts, pages, comments, and media, will be lost.'
```

File: src/components/user-settings.tsx [hunk 4]
```diff
@@ -162,11 +162,27 @@ export default function UserSettings() {
...
+			),
+			buttons: [ __( 'Cancel' ), __( 'Delete all' ) ],
+			cancelId: CANCEL_BUTTON_INDEX,
+		} );
+
+		if ( response === DELETE_BUTTON_INDEX ) {
+			await deleteAllSnapshots( allSnapshots );
```
Comment (wojtekn) [inline src/components/user-settings.tsx:194]: What do you think about confirming the action to the user in a similar way we use when the user updates the demo site?

**Comment by wojtekn** [comment]
Thanks @kozer , looks good:

Screenshot 2024-05-21 at 11 25 20

Screenshot 2024-05-21 at 11 22 28

Could you please fix lint errors?

**Review by wojtekn — APPROVED** [review]

**Pull request merged**